java/arcs/core/util/StateMachine.kt

```diff
@@ -0,0 +1,209 @@
+ *
+ * @param STATE usually an enum or sealed class representing possible states
+ * @param EVENT usually an enum or sealed class representing possible event triggered transitions
+ * @param MACHINE the subclass type that extends [StateMachine]
+ * @property state the current/starting state of the machine.
+ */
+abstract class StateMachine<STATE, EVENT, MACHINE : StateMachine<STATE, EVENT, MACHINE>>(
+    var state: STATE
+) {
+    private lateinit var states: Map<STATE, StateNode<STATE, EVENT, MACHINE>>
+    private val eventQueue = mutableListOf<EVENT>()
+
+    /** True if there are no pending events to process. */
+    val idle = eventQueue.isEmpty()
+
+    internal data class Transition<EVENT, STATE, MACHINE>(
+        val nextState: STATE,
+        val block: suspend (MACHINE).(STATE) -> Unit
+    )
+
+    internal data class StateNode<STATE, EVENT, MACHINE>(
+        val entryHandler: EntryHandler<MACHINE>,
+        val exitHandler: EntryHandler<MACHINE>,
+        val errorHandler: ErrorHandler<MACHINE>,
+        val eventHandlers: Map<EVENT, Transition<EVENT, STATE, MACHINE>>
+    )
+
+    /**
+     * Triggers an event to be queued and processed immediately. This should only be called from
+     * external code, never from state machine handlers.
+     */
+    suspend fun triggerImmediate(event: EVENT) {
+        trigger(event)
+        triggerInternal()
+    }
+
+    /**
+     * Called by state machine handlers to queue an event to be processed after processing the
+     * current event is finished processing.
+     */
+    protected suspend fun trigger(event: EVENT) {
+        eventQueue += event
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun triggerInternal() {
+        if (eventQueue.isEmpty()) {
+            return
+        }
+        val event = eventQueue.removeAt(0)
+        val stateNode = requireNotNull(states[state])
+        val eventHandler = requireNotNull(stateNode.eventHandlers[event]) {
+            "Illegal state transition $event while in state $state"
+        }
+
+        try {
+            stateNode.exitHandler.invoke(this as MACHINE)
+            eventHandler.block(this, eventHandler.nextState)
+            transitionTo(eventHandler.nextState)
+        } catch (e: Exception) {
+            eventQueue.clear()
+            stateNode.errorHandler.invoke(this as MACHINE, e)
+        }
+
+        triggerInternal()
+    }
+
+    /**
+     * Invoked when transitioning to a new state. Verifies this is a valid transition, invokes
+     * an entry handlers, forwarding exceptions to the error handlers.
+     */
+    @Suppress("UNCHECKED_CAST")
+    protected suspend fun transitionTo(newState: STATE) {
+        val destination = requireNotNull(states[newState]) {
+            "$newState is not defined in the state machine."
+        }
+
+        state = newState
+        try {
+            destination.entryHandler.invoke(this as MACHINE)
+        } catch (e: Exception) {
```

Comment:
Any exception? 

A comment explaining why would be good here, and if it's necessary - maybe `Throwable` is better?